server/src/main/java/org/elasticsearch/common/collect/Iterators.java

```diff
@@ -116,6 +116,29 @@ public T next() {
         }
     }
 
+    public static <T, U> Iterator<U> map(Iterator<? extends T> input, Function<T, U> fn) {
+        if (input.hasNext()) {
+            return new MapIterator<>(input, fn);
+        } else {
+            return Collections.emptyIterator();
+        }
+    }
+
+    private record MapIterator<T, U>(Iterator<? extends T> input, Function<T, U> fn) implements Iterator<U> {
```

Comment:
Does this need to be a record? identity/hashcode don't really make sense for Functions...

Reply:
I just followed IntelliJ's advice here, but yeah this doesn't really make sense does it